From c88bce079d3ced3ed8428de1961d6d1036b382b0 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 22 Aug 2022 13:55:38 +0300
Subject: [PATCH] [SW-86869] gaudi3: validate and override revokation bits

Currently, as part of fuse override function only key 4 revokation
bit is cleared. This patch add changes to validate revoke bits
value read from fuse (valid fuse values: 0, 1, 3, 7, 15) and override
if not in expected fuse values.

Change-Id: I38fbed42ab936100254ba6c8b73e8db9b9cfa83f
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index fb8e5fa..bc8faba 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -438,12 +438,26 @@
 
 void arcboot_rom_fuse_override(void)
 {
+	uint32_t key_revoke;
 
 	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_EOM_MASK)
 		arcboot_rom_global.config_fuse |= EFUSE_SYSCFG_PRE_PRODUCTION_BITS;
 
 	/* KEY index 4 must not be revoked */
 	arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_REVOKE_KEY_4;
+
+	/*
+	 * Validate key revokation bits. Expected values from fuse are
+	 * 0, 1, 3, 7, 15. If any other values are read from fuse override
+	 * them accordingly.
+	 */
+	key_revoke = arcboot_rom_global.config_fuse & EFUSE_SYSCFG_REVOKE_KEY_MASK;
+	if (key_revoke == 2)
+		arcboot_rom_global.config_fuse |=  EFUSE_SYSCFG_REVOKE_KEY_0_1;
+	else if ((key_revoke < 7) && (key_revoke > 3))
+		arcboot_rom_global.config_fuse |=  EFUSE_SYSCFG_REVOKE_KEY_0_1_2;
+	else if ((key_revoke < 15) && (key_revoke > 7))
+		arcboot_rom_global.config_fuse |=  EFUSE_SYSCFG_REVOKE_KEY_0_1_2_3;
 }
 
 __weak int arcboot_rom_fuse_single_die_config(void)
diff --git a/arch/gaudi3/arc/dev_efuse.h b/arch/gaudi3/arc/dev_efuse.h
index bd029c0..6792139 100644
--- a/arch/gaudi3/arc/dev_efuse.h
+++ b/arch/gaudi3/arc/dev_efuse.h
@@ -13,6 +13,9 @@
 
 #define EFUSE_SYSCFG_REVOKE_KEY_0			BIT(0)
 #define EFUSE_SYSCFG_REVOKE_KEY_4			BIT(4)
+#define EFUSE_SYSCFG_REVOKE_KEY_0_1			(BIT(0) | BIT(1))
+#define EFUSE_SYSCFG_REVOKE_KEY_0_1_2			(BIT(0) | BIT(1) | BIT(2))
+#define EFUSE_SYSCFG_REVOKE_KEY_0_1_2_3			(BIT(0) | BIT(1) | BIT(2) | BIT(3))
 #define EFUSE_SYSCFG_EOM_MASK				BIT(5)
 #define EFUSE_SYSCFG_TPM_CODE_LOCATION_MASK		BIT(6)
 #define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK	BIT(7)
